subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/AbstractCompatibilityTestRunner.java

```diff
@@ -50,40 +50,76 @@ protected AbstractCompatibilityTestRunner(Class<?> target) {
     private AbstractCompatibilityTestRunner(Class<?> target, String versionStr) {
         super(target);
         validateTestName(target);
+    }
 
-        previous = new ArrayList<GradleDistribution>();
-        final ReleasedVersionDistributions releasedVersions = new ReleasedVersionDistributions(buildContext);
-        if (versionStr.equals("latest")) {
-            implicitVersion = true;
-            addVersionIfCompatibleWithJvmAndOs(releasedVersions.getMostRecentRelease());
-        } else if (versionStr.equals("all")) {
-            implicitVersion = true;
-            List<GradleDistribution> previousVersionsToTest = choosePreviousVersionsToTest(releasedVersions);
-            for (GradleDistribution previousVersion : previousVersionsToTest) {
-                addVersionIfCompatibleWithJvmAndOs(previousVersion);
-            }
-        } else if (versionStr.matches("^\\d.*$")) {
-            implicitVersion = false;
-            String[] versions = versionStr.split(",");
-            List<GradleVersion> gradleVersions = CollectionUtils.sort(collect(Arrays.asList(versions), new Transformer<GradleVersion, String>() {
-                public GradleVersion transform(String versionString) {
-                    return GradleVersion.version(versionString);
+    abstract protected void createConfiguredExecutions();
+
+    @Override
+    protected void createExecutionsForContext(CoverageContext context) {
+        switch(context) {
```

Comment:
I'd make this more structured, as there are really only 2 patterns for a runner to choose from and we want to limit the number of patterns as much as practical and get rid of the duplication and odd inconsistencies that come from the duplication.

I would structure the runners like this:

For the base runner:

- Add a type parameter T to represent the kind of thing that the test iterates over, e.g. Gradle distributions, (tapi client, Gradle), native toolchains, Groovy compilers, whatever.
- Add an abstract method that returns all of the T that could be applicable to the test, regardless of whether that T works/is available for the current JVM/OS
- Add an abstract method to the runner (or to T) that determines whether a given T is available on the machine.
- Add a way to order T, possibly by just having the first method return the T instances in preferred order.
- In the base runner, use these methods above to decide which instances of T to run based on the context. 
- Maybe add a method to allow the runner to decide which instances of T to use for a partial test, given a list of T that are _available_ in reverse version order, so that a specific runner can choose whether to use just the latest or latest and oldest. Maybe also the same for a quick test.
- Get rid of `createExecutionsForContext()` and `createSelectedExecutions()` as the base runner can use the above methods to implement these.